effectful/src/Effectful/FileSystem.hs

```diff
@@ -0,0 +1,357 @@
+
+      -- * Timestamps
+    , getAccessTime
+    , getModificationTime
+    , setAccessTime
+    , setModificationTime
+
+      -- * Re-exports
+
+      -- ** Pre-defined directories
+    , D.XdgDirectory(..)
+    , D.XdgDirectoryList(..)
+
+      -- ** Existence tests
+    , D.exeExtension
+
+      -- ** Permissions
+    , D.Permissions
+    , D.emptyPermissions
+    , D.readable
+    , D.writable
+    , D.executable
+    , D.searchable
+    , D.setOwnerReadable
+    , D.setOwnerWritable
+    , D.setOwnerExecutable
+    , D.setOwnerSearchable
+    ) where
+
+import Data.Time (UTCTime)
+import qualified System.Directory as D
+
+import Effectful.Internal.Effect
+import Effectful.Internal.Monad
+
+-- | An effect for interacting with the filesystem.
+data FileSystem :: Effect where
+  FileSystem :: FileSystem m r
+
+-- | Run the 'FileSystem' effect.
+runFileSystem :: IOE :> es => Eff (FileSystem : es) a -> Eff es a
+runFileSystem = evalEffect (IdE FileSystem)
+
+----------------------------------------
+-- Actions on directories
+
+-- | Lifted 'D.createDirectory'.
+createDirectory :: FileSystem :> es => FilePath -> Eff es ()
+createDirectory = unsafeEff_ . D.createDirectory
+
+-- | Lifted 'D.createDirectoryIfMissing'.
+createDirectoryIfMissing :: FileSystem :> es => Bool -> FilePath -> Eff es ()
+createDirectoryIfMissing doCreateParents =
+  unsafeEff_ . D.createDirectoryIfMissing doCreateParents
+
+-- | Lifted 'D.removeDirectory'.
+removeDirectory :: FileSystem :> es => FilePath -> Eff es ()
+removeDirectory = unsafeEff_ . D.removeDirectory
+
+-- | Lifted 'D.removeDirectoryRecursive'.
+removeDirectoryRecursive :: FileSystem :> es => FilePath -> Eff es ()
+removeDirectoryRecursive = unsafeEff_ . D.removeDirectoryRecursive
+
+-- | Lifted 'D.removePathForcibly'.
+removePathForcibly :: FileSystem :> es => FilePath -> Eff es ()
+removePathForcibly = unsafeEff_ . D.removePathForcibly
+
+-- | Lifted 'D.renameDirectory'.
+renameDirectory :: FileSystem :> es => FilePath -> FilePath -> Eff es ()
+renameDirectory old = unsafeEff_ . D.renameDirectory old
+
+-- | Lifted 'D.listDirectory'.
+listDirectory :: FileSystem :> es => FilePath -> Eff es [FilePath]
+listDirectory = unsafeEff_ . D.listDirectory
+
+-- | Lifted 'D.getDirectoryContents'.
+getDirectoryContents :: FileSystem :> es => FilePath -> Eff es [FilePath]
+getDirectoryContents = unsafeEff_ . D.getDirectoryContents
+
+----------------------------------------
+-- Current working directory
+
+-- | Lifted 'D.getCurrentDirectory'.
+getCurrentDirectory :: FileSystem :> es => Eff es FilePath
+getCurrentDirectory = unsafeEff_ D.getCurrentDirectory
+
+-- | Lifted 'D.setCurrentDirectory'.
+setCurrentDirectory :: FileSystem :> es => FilePath -> Eff es ()
+setCurrentDirectory = unsafeEff_ . D.setCurrentDirectory
+
+-- | Lifted 'D.withCurrentDirectory'.
+withCurrentDirectory :: FileSystem :> es => FilePath -> Eff es a -> Eff es a
+withCurrentDirectory path k = unsafeEff $ \es -> do
+  seqUnliftEff es $ \unlift ->
```

Comment:
You can use `unsafeLiftMapIO` here, it'll be a bit simpler. I think it's fine to assume that these functions will not use multiple threads. :thinking: Same below.

Reply:
Done.